File: implementations/typescript/packages/llmproxy/src/mcp-refresh.ts
// SPDX-License-Identifier: Apache-2.0
import {canonicalVersion,parseMarkup,sectionToEnvelope,type Envelope} from "@psp-cdl/core";
import {identifier,type Principal} from "@psp-cdl/api-server";
import {bounded} from "@psp-cdl/api-server/persistence";
import type {PinnedMcpClient} from "@psp-cdl/mcpproxy/mcp";
import {refreshToolDefinition} from "./mcp-refresh-contract.js";
export const MCP_PROMPT_REFRESH_PROFILE="PSP-MCP-PROMPT-REFRESH-0.1";
export class PromptRefreshError extends Error {
  constructor(public readonly code:string){super(code);this.name="PromptRefreshError";}
}
const fail=(code:string):never=>{throw new PromptRefreshError(code);};
/** A detached definition for a host's dedicated authenticated refresh service. */
export const mcpRefreshToolDefinition=():typeof refreshToolDefinition=>bounded(refreshToolDefinition) as typeof refreshToolDefinition;

/** Bind once to an approved peer/catalog and exact owner/session, outside the model tool registry. */
export class McpPromptRefresher {
  private readonly invoke:ReturnType<PinnedMcpClient["bindControlTool"]>;
  private readonly tenantId:string;
  private readonly subjectId:string;
  private readonly sessionId:string;
  private readonly cancelled:()=>boolean;
  constructor(private readonly peer:PinnedMcpClient,config:{principal:Principal;sessionId:string;approvedCatalogDigest:string;now:()=>number;cancelled:()=>boolean;toolName?:string;toolRevision?:string}) {
    if(!config||!config.principal||![config.principal.tenantId,config.principal.subjectId,config.sessionId].every(identifier)||typeof config.now!=="function"||typeof config.cancelled!=="function")fail("INVALID_CONFIGURATION");
    this.tenantId=config.principal.tenantId;this.subjectId=config.principal.subjectId;this.sessionId=config.sessionId;this.cancelled=config.cancelled;
    const definition=mcpRefreshToolDefinition();
    this.invoke=peer.bindControlTool({name:config.toolName??definition.name,revision:config.toolRevision??"refresh-1",inputSchema:definition.inputSchema,outputSchema:definition.outputSchema},config.approvedCatalogDigest,config.now);
  }
  async refresh(principal:Principal,binding:Record<string,unknown>,request:Record<string,unknown>):Promise<Envelope> {
    if(!principal||!binding||!request||principal.tenantId!==this.tenantId||principal.subjectId!==this.subjectId||binding.tenantId!==this.tenantId||binding.subjectId!==this.subjectId||binding.sessionId!==this.sessionId||request.session_id!==this.sessionId)fail("REFRESH_BINDING_MISMATCH");
    try{if(typeof request.current_version!=="string"||canonicalVersion(request.current_version)!==request.current_version)fail("INVALID_REFRESH_REQUEST");}catch{fail("INVALID_REFRESH_REQUEST");}
    const result=await this.invoke(request,{deadline:binding.deadline as number,cancelled:this.cancelled}) as {prompt:string};
    try {
      const doc=parseMarkup(result.prompt);
      const section=doc.children[0];
      if(doc.children.length!==1||!section||section.kind!=="section")return fail("INVALID_REFRESH_RESPONSE");
      const envelope=sectionToEnvelope(section);
      if(envelope.signature.sectionType!=="system"||envelope.signature.contentType!=="text")fail("INVALID_REFRESH_RESPONSE");
      // Structural parsing is not verification. RefreshingLlmLoop checks current keys, scope, freshness and host approval.
      return envelope;
    }catch{await this.peer.close();return fail("INVALID_REFRESH_RESPONSE");}
  }
}

File: implementations/python/packages/llmproxy/src/psp_cdl_llmproxy/mcp_refresh.py
# SPDX-License-Identifier: Apache-2.0
"""Host-only MCP discovery for the opt-in prompt refresh loop."""
from copy import deepcopy
from psp_cdl_core import canonical_version, parse_markup, section_to_envelope
from psp_cdl_api_server.service import identifier
from .mcp_refresh_contract import REFRESH_TOOL_DEFINITION
MCP_PROMPT_REFRESH_PROFILE="PSP-MCP-PROMPT-REFRESH-0.1"

class PromptRefreshError(ValueError):
    def __init__(self,code):super().__init__(code);self.code=code

def mcp_refresh_tool_definition():return deepcopy(REFRESH_TOOL_DEFINITION)

class McpPromptRefresher:
    def __init__(self,peer,config):
        p=config.get("principal") if type(config) is dict else None
        if type(p) is not dict or not all(identifier(v) for v in (p.get("tenantId"),p.get("subjectId"),config.get("sessionId"))) or not callable(config.get("now")) or not callable(config.get("cancelled")):raise PromptRefreshError("INVALID_CONFIGURATION")
        self._peer,self._tenant,self._subject,self._session,self._cancelled=peer,p["tenantId"],p["subjectId"],config["sessionId"],config["cancelled"]
        tool=mcp_refresh_tool_definition()
        self._invoke=peer.bind_control_tool({"name":config.get("toolName",tool["name"]),"revision":config.get("toolRevision","refresh-1"),"inputSchema":tool["inputSchema"],"outputSchema":tool["outputSchema"]},config.get("approvedCatalogDigest"),config["now"])

    def refresh(self,principal,binding,request):
        if type(principal) is not dict or type(binding) is not dict or type(request) is not dict or any(value.get("tenantId")!=self._tenant or value.get("subjectId")!=self._subject for value in (principal,binding)) or binding.get("sessionId")!=self._session or request.get("session_id")!=self._session:raise PromptRefreshError("REFRESH_BINDING_MISMATCH")
        try:
            if type(request.get("current_version")) is not str or canonical_version(request["current_version"])!=request["current_version"]:raise ValueError()
        except Exception:raise PromptRefreshError("INVALID_REFRESH_REQUEST") from None
        result=self._invoke(request,{"deadline":binding.get("deadline"),"cancelled":self._cancelled})
        try:
            doc=parse_markup(result["prompt"])
            if len(doc["children"])!=1 or doc["children"][0]["kind"]!="section":raise ValueError()
            envelope=section_to_envelope(doc["children"][0])
            if envelope["signature"]["sectionType"]!="system" or envelope["signature"]["contentType"]!="text":raise ValueError()
            return envelope
        except Exception:
            self._peer.close();raise PromptRefreshError("INVALID_REFRESH_RESPONSE") from None
